pkg/gohai/filesystem/filesystem_nix.go

```diff
@@ -0,0 +1,196 @@
+	"vxfs":       {},
+}
+
+// FSInfoGetter provides function to get information about a given filesystem: its size and its dev id
+type FSInfoGetter interface {
+	// Size returns the size of the given filesystem
+	Size(mount *mountinfo.Info) (uint64, error)
+
+	// Dev returns the dev id of the given filesystem
+	// the return type is interface{} because `syscall.Stat_t` uses different types for Dev depending on the platform
+	Dev(mount *mountinfo.Info) (interface{}, error)
+}
+
+type UnixFSInfo struct{}
+
+func (UnixFSInfo) Size(mount *mountinfo.Info) (uint64, error) {
+	var statfs unix.Statfs_t
+	if err := unix.Statfs(mount.Mountpoint, &statfs); err != nil {
+		return 0, fmt.Errorf("statfs %s: %v", mount.Source, err)
+	}
+
+	sizeKB := statfs.Blocks * uint64(statfs.Bsize) / 1024
+	return sizeKB, nil
+}
+
+func (UnixFSInfo) Dev(mount *mountinfo.Info) (interface{}, error) {
+	var stat unix.Stat_t
+	if err := unix.Stat(mount.Mountpoint, &stat); err != nil {
+		return 0, fmt.Errorf("stat %s: %w", mount.Source, err)
+	}
+
+	return stat.Dev, nil
+}
+
+func getFileSystemInfo() ([]MountInfo, error) {
+	return getFileSystemInfoWithMounts(nil, UnixFSInfo{})
+}
+
+// returns whether to use the new mountInfo instead of the old one
+func replaceDev(old, new MountInfo) bool {
+	if strings.ContainsRune(new.Name, '/') && !strings.ContainsRune(old.Name, '/') {
+		return true
+	}
+	if len(old.MountedOn) > len(new.MountedOn) {
+		return true
+	}
+
+	return old.Name != new.Name && old.MountedOn == new.MountedOn
+}
+
+// Internal method to help testing with test mounts
+func getFileSystemInfoWithMounts(initialMounts []*mountinfo.Info, fsInfo FSInfoGetter) ([]MountInfo, error) {
+	var err error
+	mounts := initialMounts
+
+	if mounts == nil {
+		mounts, err = mountinfo.GetMounts(nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	devMountInfos := map[interface{}]MountInfo{}
+	for _, mount := range mounts {
+		// Skip mounts that seem to be missing data
+		if mount.Source == "" || mount.Source == "none" || mount.FSType == "" || mount.Mountpoint == "" {
+			continue
+		}
+
+		if isExcludedFS(mount.FSType, mount.Source, true) {
+			continue
+		}
+
+		sizeKB, err := fsInfo.Size(mount)
+		if err != nil {
+			log.Info(err)
+			continue
+		}
+
+		if sizeKB == 0 {
+			// ignore zero-sized filesystems, like `df` does (unless using -a)
+			continue
+		}
+
+		mountInfo := MountInfo{
+			Name:      mount.Source,
+			SizeKB:    sizeKB,
+			MountedOn: mount.Mountpoint,
+		}
+
+		dev, err := fsInfo.Dev(mount)
+		if err != nil {
+			log.Info(err)
+			continue
+		}
+
+		existingMountInfo, exists := devMountInfos[dev]
+		if !exists || replaceDev(existingMountInfo, mountInfo) {
+			devMountInfos[dev] = mountInfo
+		}
+	}
+
+	mountInfos := make([]MountInfo, 0, len(devMountInfos))
+	for _, mountInfo := range devMountInfos {
+		mountInfos = append(mountInfos, mountInfo)
+	}
+
+	return mountInfos, nil
+}
+
+func isExcludedFS(fsType string, fsSource string, localOnly bool) bool {
```

Comment:
I thought about it (it was implemented like that in the original [PR](https://github.com/DataDog/gohai/pull/129)) and initially considered that we might as well keep it in case we decide to keep remote fs at some point, but we could still implement it back if needed.